### PR TITLE
gh-117709: Add vectorcall support for positional-only arguments of str()

### PR DESCRIPTION
**Pull request by erlend-aasland**
Competing PR to gh-117725.

* Issue: gh-117709

### CODE REVIEW

File: Misc/NEWS.d/next/Core and Builtins/2024-04-10-22-16-18.gh-issue-117709.-_1YL0.rst [hunk 1]
```diff
@@ -0,0 +1,2 @@
+Speed up calls to :func:`str` by using the :pep:`590` ``vectorcall`` calling
+convention. Patch by Erlend Aasland.
```
Comment (erlend-aasland) [inline Misc/NEWS.d/next/Core and Builtins/2024-04-10-22-16-18.gh-issue-117709.-_1YL0.rst]:
```suggestion
Speed up calls to :func:`str` with positional-only argument,
by using the :pep:`590` ``vectorcall`` calling convention.
Patch by Erlend Aasland.
```

File: Objects/unicodeobject.c [hunk 2]
```diff
@@ -14617,6 +14617,67 @@ unicode_new_impl(PyTypeObject *type, PyObject *x, const char *encoding,
...
+    const char *encoding = NULL;
+    const char *errors = NULL;
+    if (nargs == 2) {
+        encoding = as_const_char(args[1], "encoding");
+    }
+    if (nargs == 3) {
+        encoding = as_const_char(args[1], "encoding");
+        errors = as_const_char(args[1], "errors");
+    }
+    return PyUnicode_FromEncodedObject(object, encoding, errors);
```
Comment (erlend-aasland) [inline Objects/unicodeobject.c]:
```suggestion
    const char *encoding = as_const_char(args[1], "encoding");
    const char *errors = (nargs == 2) ? NULL : as_const_char(args[2], "errors");
    return PyUnicode_FromEncodedObject(object, encoding, errors);
```

**Comment by erlend-aasland** [comment]
The speedups are as with #117725, but only for the positional-only cases.

TL;DR: no change in performance for the single-argument case nor the fallback to `tp_call`; approx. 2x increase in performance for two or three positional-only arguments[^1].


[^1]: release build on macOS.

File: Objects/unicodeobject.c [hunk 3]
```diff
@@ -14617,6 +14617,60 @@ unicode_new_impl(PyTypeObject *type, PyObject *x, const char *encoding,
...
+        PyObject *dict = _PyStack_AsDict(args + nargs, kwnames);
+        if (dict == NULL) {
+            Py_DECREF(tuple);
+            return NULL;
+        }
+        return unicode_new(_PyType_CAST(type), tuple, dict);
+    }
+    if (nargs > 3) {
+        PyErr_Format(PyExc_TypeError,
+            "str() takes at most 3 arguments (%d given)", nargs + nkwargs);
```
Comment (vstinner) [inline Objects/unicodeobject.c]:
```suggestion
            "str() takes at most 3 arguments (%d given)", nargs);
```

File: Objects/unicodeobject.c [hunk 4]
```diff
@@ -14617,6 +14617,60 @@ unicode_new_impl(PyTypeObject *type, PyObject *x, const char *encoding,
...
+
+static PyObject *
+unicode_vectorcall(PyObject *type, PyObject *const *args,
+                   size_t nargsf, PyObject *kwnames)
+{
+    assert(Py_Is(_PyType_CAST(type), &PyUnicode_Type));
+
+    Py_ssize_t nargs = PyVectorcall_NARGS(nargsf);
+    Py_ssize_t nkwargs = (kwnames) ? PyTuple_GET_SIZE(kwnames) : 0;
+    if (nkwargs) {
```
Comment (vstinner) [inline Objects/unicodeobject.c]:
```suggestion
    if (kwnames != NULL && PyTuple_GET_SIZE(kwnames) != 0) {
```

File: Objects/unicodeobject.c [hunk 5]
```diff
@@ -14617,6 +14617,60 @@ unicode_new_impl(PyTypeObject *type, PyObject *x, const char *encoding,
...
+        // Fallback to tp_call()
```
Comment (vstinner) [inline Objects/unicodeobject.c]:
```suggestion
        // Fallback to unicode_new()
```

File: Objects/unicodeobject.c [hunk 6]
```diff
@@ -14617,6 +14617,60 @@ unicode_new_impl(PyTypeObject *type, PyObject *x, const char *encoding,
...
+        PyObject *tuple = _PyTuple_FromArray(args, nargs);
+        if (tuple == NULL) {
+            return NULL;
+        }
+        PyObject *dict = _PyStack_AsDict(args + nargs, kwnames);
+        if (dict == NULL) {
+            Py_DECREF(tuple);
+            return NULL;
+        }
+        return unicode_new(_PyType_CAST(type), tuple, dict);
```
Comment (vstinner) [inline Objects/unicodeobject.c]: You have to DECREF tuple and dict here, no?

File: Objects/unicodeobject.c [hunk 7]
```diff
@@ -14617,6 +14617,60 @@ unicode_new_impl(PyTypeObject *type, PyObject *x, const char *encoding,
     return unicode;
 }
 
+static const char *
+as_const_char(PyObject *obj, const char *name)
```
Comment (vstinner) [inline Objects/unicodeobject.c]: I suggest to rename the function to `arg_as_utf8()`.

**Review by vstinner — COMMENTED** [review]
I prefer this PR over https://github.com/python/cpython/pull/117725 which is way more complex.

File: Objects/unicodeobject.c [hunk 8]
```diff
@@ -14617,6 +14617,60 @@ unicode_new_impl(PyTypeObject *type, PyObject *x, const char *encoding,
...
+                     name, obj);
+        return NULL;
+    }
+    const char *str = _PyUnicode_AsUTF8NoNUL(obj);
+    if (str == NULL) {
+        return NULL;
+    }
+    return str;
+}
+
```
Comment (vstinner) [inline Objects/unicodeobject.c:14631]: You may add a comment to mention that this function is a fast-path for positional-only cases.

File: Objects/unicodeobject.c (hunk 6 above)
Comment (erlend-aasland) [reply Objects/unicodeobject.c]: Yes, absolutely!

**Review by corona10 — COMMENTED** [review]
I personally prefer this one :)
Do we have enough tests for positional arguments cases?

**Comment by erlend-aasland** [comment]
> I personally prefer this one :) Do we have enough tests for positional arguments cases?

Well, we could add something like this:

<details>

```diff
diff --git a/Lib/test/test_str.py b/Lib/test/test_str.py
index b4927113db..ea37eb5d96 100644
--- a/Lib/test/test_str.py
+++ b/Lib/test/test_str.py
@@ -2651,6 +2651,24 @@ def test_check_encoding_errors(self):
         proc = assert_python_failure('-X', 'dev', '-c', code)
         self.assertEqual(proc.rc, 10, proc)
 
+    def test_str_invalid_call(self):
+        check = lambda *a, **kw: self.assertRaises(TypeError, str, *a, **kw)
+
+        # too many args
+        check(1, "", "", 1)
+
+        # no such kw arg
+        check(test=1)
+
+        # 'encoding' must be str
+        check(1, encoding=1)
+        check(1, 1)
+
+        # 'errors' must be str
+        check(1, errors=1)
+        check(1, "", errors=1)
+        check(1, 1, 1)
+
 
 class StringModuleTest(unittest.TestCase):
     def test_formatter_parser(self):
```

</details>

File: Objects/unicodeobject.c [hunk 9]
```diff
@@ -14617,6 +14617,62 @@ unicode_new_impl(PyTypeObject *type, PyObject *x, const char *encoding,
...
+        if (dict == NULL) {
+            Py_DECREF(tuple);
+            return NULL;
+        }
+        PyObject *ret = unicode_new(_PyType_CAST(type), tuple, dict);
+        Py_DECREF(tuple);
+        Py_DECREF(dict);
+        return ret;
+    }
+    if (nargs > 3) {
```
Comment (corona10) [inline Objects/unicodeobject.c]: Can we use ths one?

```suggestion
    if (!_PyArg_CheckPositional("str", nargs, 0, 3)) {
        return NULL;
    }
```

File: Objects/unicodeobject.c [hunk 10]
```diff
@@ -14617,6 +14617,62 @@ unicode_new_impl(PyTypeObject *type, PyObject *x, const char *encoding,
...
+    }
+    if (nargs == 0) {
+        return unicode_get_empty();
+    }
+    PyObject *object = args[0];
+    if (nargs == 1) {
+        return PyObject_Str(object);
+    }
+    const char *encoding = arg_as_utf8(args[1], "encoding");
+    const char *errors = (nargs == 2) ? NULL : arg_as_utf8(args[2], "errors");
```
Comment (erlend-aasland) [inline Objects/unicodeobject.c]: This reads better:

```suggestion
    const char *errors = (nargs == 3) ? arg_as_utf8(args[2], "errors") : NULL;
```

File: Objects/unicodeobject.c (hunk 9 above)
Comment (erlend-aasland) [reply Objects/unicodeobject.c]: Well, it's more expensive; moreover there is precedent for this variant in the code base.

Comment (erlend-aasland) [reply Objects/unicodeobject.c]: I would not oppose it though.

File: Objects/unicodeobject.c [hunk 11]
```diff
@@ -14617,6 +14617,62 @@ unicode_new_impl(PyTypeObject *type, PyObject *x, const char *encoding,
...
+        PyObject *ret = unicode_new(_PyType_CAST(type), tuple, dict);
+        Py_DECREF(tuple);
+        Py_DECREF(dict);
+        return ret;
+    }
+    if (nargs > 3) {
+        PyErr_Format(PyExc_TypeError,
+            "str() takes at most 3 arguments (%d given)", nargs);
+        return NULL;
+    }
```
Comment (erlend-aasland) [inline Objects/unicodeobject.c]: @corona10: this would be the actual change.

```suggestion
    if (!_PyArg_CheckPositional("str", nargs, 0, 3)) {
        return NULL;
    }
```

File: Objects/unicodeobject.c (hunk 9 above)
Comment (erlend-aasland) [reply Objects/unicodeobject.c]: See https://github.com/python/cpython/pull/117746/files#r1560988495 for the actual suggestion.

File: Objects/unicodeobject.c (hunk 11 above)
Comment (corona10) [reply Objects/unicodeobject.c]: Yeah, yeah, I was lazy :)

Comment (erlend-aasland) [reply Objects/unicodeobject.c]: Note that this would change the exception message. Currently we have this:

```console
$ python3
Python 3.13.0a6 (v3.13.0a6:57aee2a02c, Apr  9 2024, 09:56:43) [Clang 13.0.0 (clang-1300.0.29.30)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> str(1, 1, 1, 1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
    str(1, 1, 1, 1)
    ~~~^^^^^^^^^^^^
TypeError: str() takes at most 3 arguments (4 given)
```

With `_PyArg_CheckPositional` we get this:
```console
$ ./python.exe
Python 3.13.0a6+ (heads/perf/unicode-vectorcall-pos-only-dirty:8138db40dd, Apr 11 2024, 15:03:45) [Clang 15.0.0 (clang-1500.3.9.4)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> str(1, 1, 1, 1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
    str(1, 1, 1, 1)
    ~~~^^^^^^^^^^^^
TypeError: str expected at most 3 arguments, got 4
```

Perhaps it does not matter.

Comment (corona10) [reply Objects/unicodeobject.c]: FYI, All vectorcall implementations are using _PyArg_CheckPositional as possible.
(To reduce redundant code)

Comment (erlend-aasland) [reply Objects/unicodeobject.c]: @corona10, not `enumerate` ;)

Comment (corona10) [reply Objects/unicodeobject.c]: > @corona10, not enumerate ;)

Yeah, there are some exceptional case, but please don't generalize it :)

> Perhaps it does not matter.

Yeah, We've changed the same things for other objects.

Comment (erlend-aasland) [reply Objects/unicodeobject.c]: oh, also not `long`.

Comment (erlend-aasland) [reply Objects/unicodeobject.c]: ... nor `range` :)

Comment (erlend-aasland) [reply Objects/unicodeobject.c]: But sure, yes, the majority of them use `_PyArg_CheckPositional`.

**Review by corona10 — APPROVED** [review]
lgtm

File: Objects/unicodeobject.c [hunk 12]
```diff
@@ -14617,6 +14617,60 @@ unicode_new_impl(PyTypeObject *type, PyObject *x, const char *encoding,
...
+        PyErr_Format(PyExc_TypeError,
+                     "str() argument '%s' must be str, not %T",
+                     name, obj);
+        return NULL;
+    }
+    const char *str = _PyUnicode_AsUTF8NoNUL(obj);
+    if (str == NULL) {
+        return NULL;
+    }
+    return str;
```
Comment (vstinner) [inline Objects/unicodeobject.c]:
```suggestion
    return _PyUnicode_AsUTF8NoNUL(obj);
```

**Review by vstinner — APPROVED** [review]
LGTM. I added a suggestion.

I'm not sure that arg_as_utf8() helper is strictly needed, you can keep it or inline it, it's up to you ;-)

File: Objects/unicodeobject.c (hunk 12 above)
Comment (erlend-aasland) [reply Objects/unicodeobject.c]: Thanks for pointing out the obvious 😁

**Comment by erlend-aasland** [comment]
Thank you so much for the reviews, Donghee and Victor!
